javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala

```diff
@@ -0,0 +1,244 @@
+package java.util.concurrent
+
+import java.util._
+import scala.collection.JavaConversions._
+
+trait ConcurrentMap[K, V] extends java.util.Map[K, V] {
+
+  def remove(key: Any, value: Any): Boolean = {
+    if (containsKey(key) && get(key) === value) {
+      remove(key)
+      true
+    } else false
+  }
+
+  def replace(key: K, value: V): V = {
+    if (containsKey(key)) {
+      put(key, value)
+    } else get(key)
+  }
+
+  def replace(key: K, oldValue: V, newValue: V): Boolean = {
+    if (containsKey(key) && get(key) === oldValue) {
+      put(key, newValue)
+      true
+    } else false
+  }
+}
+
+class ConcurrentHashMap[K >: Null, V >: Null](initialCapacity: Int,
+    loadFactor: Float,
+    concurrencyLevel: Int)
+    extends java.util.AbstractMap[K, V]
+    with java.util.concurrent.ConcurrentMap[K, V]
+    with java.io.Serializable { self =>
+
+  def this() = 
+    this(16, 0.75F, 16)
+
+  def this(initialCapacity: Int) = 
+    this(initialCapacity, 0.75F, 16)
+
+  def this(initialCapacity: Int, loadFactor: Float) = 
+    this(initialCapacity, loadFactor, 16)
+
+  def this(initialMap: java.util.Map[_ <: K, _ <: V]) = {
+    this(16, 0.75F, 16)
+    putAll(initialMap)
+  }
+
+  private final case class Box[K](inner: K) {
```

Comment:
Ah nice ^^

I think you can make `K` covariant, here: `+K`. Not sure it helps any where in particular, but it just "feels right" to do so.

Also, unless I'm mistaken, this does not rely on anything from the enclosing CHM, so you can put it in the companion object (which is better for case classes, I always feel bad when I see a case class inside a class).